### PR TITLE
feat(cfbd): add the passing and rushing endpoint families (CFBD coverage 74/84 -> 84/84)

### PR DESCRIPTION
**Pull request by saiemgilani**
Audited `cfbfastR` and `hoopR` against the live CollegeFootballData and CollegeBasketballData specs, then closed the gaps.

## Coverage

| package | API | before | after |
|---|---|---|---|
| **hoopR** | CBBD v1.28.0 | **38 / 38** | no change needed |
| **cfbfastR** | CFBD v5.27.1 | 74 / 84 | **84 / 84** |

hoopR routes everything through `.cbbd_get()` and was already complete; the parameterised paths (`/lineups/game/{gameId}`, `/substitutions/player/{playerId}`) were verified individually rather than by regex.

### Getting the spec, for whoever re-audits next

The OpenAPI document is **not** at `/swagger/v1/swagger.json` — that serves the Swagger UI shell, and `swagger-initializer.js` still points at the petstore demo. The real document is embedded in `https://api.collegefootballdata.com/swagger/swagger-ui-init.js` as `options.swaggerDoc`. (`sdv-swagger/cfbd-swagger.json` is that same captured object, which is why it has a `{"swaggerDoc": ...};` wrapper and does not parse as JSON. It is also stale — 2026-06-04.)

A coverage diff must also match **both** URL shapes used in `R/`: nearly every wrapper uses a full `https://api.collegefootballdata.com/...` literal, but `cfbd_metrics_fg_ep()` alone builds from `endpoint_path <- "metrics/fg/ep"`. A first pass grepping only the literal reported it as an eleventh gap; it is implemented.

## What this adds

| function | endpoint |
|---|---|
| `cfbd_passing_players_season()` | `/passing/players/season` |
| `cfbd_passing_players_games()` | `/passing/players/games` |
| `cfbd_passing_teams_season()` | `/passing/teams/season` |
| `cfbd_passing_teams_games()` | `/passing/teams/games` |
| `cfbd_passing_plays()` | `/passing/plays` |
| `cfbd_rushing_players_season()` | `/rushing/players/season` |
| `cfbd_rushing_players_games()` | `/rushing/players/games` |
| `cfbd_rushing_teams_season()` | `/rushing/teams/season` |
| `cfbd_rushing_teams_games()` | `/rushing/teams/games` |
| `cfbd_rushing_plays()` | `/rushing/plays` |

These carry detail the `cfbd_stats_*` and `cfbd_metrics_*` families do not: passing production split across seven pass locations, rushing across four run directions, and play-level frames with target ids, run direction and carrier attribution.

## Why the frames are 371 columns

Structural, not accidental. From the spec: `PassingProduction` is 23 stats plus `locations`; `PassingLocations` is exactly seven buckets each holding those same 23 stats; `TeamPassingSeason` is 3 identity + `offense` + `defense`. So `cfbd_passing_teams_season()` is `3 + 2 x (23 + 7 x 23)` = **371**, exactly as observed. Rushing is the same shape with `directions` and a 24/26-column block over four buckets.

cfbfastR's largest existing returns table is 82 vars. A 371-row table would be worse documentation than the shape itself, so `?cfbd_passing` and `?cfbd_rushing` document the production block and the bucket naming scheme **once** and every column follows from it. `/passing/plays` (39) and `/rushing/plays` (34) get ordinary full tables.

Worth noting for anyone reading the flattened frames: bucket names collide with stat names (`total`, `successful`), so inferring the buckets from column names is unreliable — an early attempt produced 21 "locations" instead of 7. The schema is authoritative.

## Three things established by probing, not assumed

- **2025 onward only.** 2015, 2018, 2021, 2023 and 2024 all answer HTTP 200 with `[]`; 2025 returns 819/1622 rows. `min_year_map_df` records 2025 (99 → 109 rows).
- **Empty responses need a guard.** That `[]` parses to a zero-length list, and `janitor::clean_names()` then aborts on the absent dimnames, so the caller sees a parse error instead of "no rows". All ten guard it. *(`cfbd_metrics_wepa_*` has the identical unguarded behaviour — left alone as pre-existing and out of scope, but noted in `CLAUDE.md` and the copilot instructions.)* Because these endpoints only cover 2025, the empty path is the common case here, not an edge.
- **Several play-level columns are unpopulated upstream.** `air_yards` is `NA` for **0 of 2,003** week-5 completions, as are `pass_depth`/`pass_direction`/`pass_location` and `rush_direction` — while `target` is populated (1,976 of 2,003) and the *season* aggregates do carry air yards. Documented as observed rather than as the schema intends.

## The `*_available` trap

CFBD parses air yards, YAC, location and direction out of play text, so an attempt can be counted while those stay unknown. `*_available` columns are the **denominators**, not statistics — dividing by `attempts` understates every mean. The vignette shows it concretely: Arch Manning's 2025 aDOT is **8.54** against `air_yards_attempts_available` and **4.52** against `attempts`, because 189 of his 399 attempts have no parsed location.

## Surfaces updated

`R/cfbd_passing.R` · `R/cfbd_rushing.R` · 10 test files · 12 man pages · `NAMESPACE` · `_pkgdown.yml` (family index + two new subtitles) · `NEWS.md` · `cran-comments.md` · `data/min_year_map_df.rda` · `CLAUDE.md` · `.github/copilot-instructions.md` · `README.Rmd`/`README.md` · `vignettes/cfbd_stats.Rmd`.

## Testing

- All ten functions exercised live with exact column-count assertions (189/193/371/375/39 and 89/93/175/179/34).
- Tests assert identity columns plus structural invariants (bucket counts, `offense_`/`defense_` split, column totals) rather than 371 literal names, and cover the out-of-coverage empty return and enum validation. 52 assertions, all passing.
- **`R CMD check`: Status OK** — 0 errors, 0 warnings, 0 notes, including `--run-donttest` against the live API.
- Documented with roxygen2 **8.1.0** to match `Config/roxygen2/version`; a local 7.3.3 silently rewrites all 162 man pages into an older cross-reference format, so this PR touches only the 12 new ones.

## Summary by Sourcery

Complete CFBD API coverage by adding passing and rushing endpoint families with robust empty-response handling, validation, documentation, and live API tests.

New Features:
- Add ten CollegeFootballData passing and rushing wrappers covering player, team, season, game, and play-level endpoints.
- Expose detailed passing-location and rushing-direction production data alongside play-level attribution and analysis fields.

Bug Fixes:
- Return empty data frames for out-of-coverage passing and rushing seasons instead of reporting parsing errors.
- Improve year validation so missing and NA values produce clear validation errors.

Enhancements:
- Document the wide passing and rushing return shapes, available-value denominators, season coverage, and observed upstream parsing limitations.
- Organize the new endpoint families in package references and provide CFBD coverage guidance.

Documentation:
- Update the README, NEWS, vignette, package reference, and generated man pages for the new CFBD endpoint families and their data caveats.

Tests:
- Add live API tests covering endpoint schemas, structural column invariants, filters, empty responses, enum validation, required years, and year validation.

Chores:
- Update contributor and project guidance with CFBD endpoint auditing and passing/rushing data conventions.

## Summary by CodeRabbit

- **New Features**
  - Added passing and rushing data for players, teams, seasons, games, and individual plays.
  - Expanded CollegeFootballData coverage to all 84 documented endpoints.
  - Added filters for teams, players, weeks, plays, outcomes, locations, and run directions.
  - Passing and rushing data are available from 2025 onward; earlier seasons return empty results.

- **Documentation**
  - Added usage examples, field descriptions, split metrics, and denominator guidance.

- **Tests**
  - Added coverage for response formats, filters, validation, and earlier-season behavior.

### CODE REVIEW

File: man/cfbd_rushing_teams_season.Rd [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+\value{
+\code{\link[=cfbd_rushing_teams_season]{cfbd_rushing_teams_season()}} - A data frame with 175 variables:\tabular{lll}{
+   col_name \tab types \tab description \cr
+   season \tab integer \tab Four-digit season year (e.g. 2025). \cr
+   team \tab character \tab Team name. \cr
+   conference \tab character \tab Team conference name. \cr
+}
+
+
+plus TWO copies of the 26-column production block and its four \verb{directions_*} repeats: \verb{offense_*} (the team's own rushing) and \verb{defense_*} (rushing allowed).
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply man/cfbd_rushing_teams_season.Rd]: Fixed in `d62438f8`. You are right and the arithmetic was the tell: read literally my sentence gives 3 + 2 x (26 + 4 x 26) = 263, not 175. The direction buckets are 15 columns wide, which the tests already asserted (`4L * 15L`) while the prose did not say so.

Both team topics now spell it out — `3 + 2 x (26 + 4 x 15) = 175` for rushing and `3 + 2 x (23 + 7 x 23) = 371` for passing — and `cfbd_rushing_teams_games()` carries the same for 179. Changed in the roxygen in `R/cfbd_rushing.R` and regenerated, not in `man/`.

File: R/cfbd_passing.R [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+  # Validation Lists ----
+  outcomes <- c("completion", "incompletion", "interception")
+
+  # Validation ----
+  validate_api_key()
+  validate_year(year)
+  validate_week(week)
+  validate_id(game_id)
+  if (!is.null(season_type)) validate_season_type(season_type)
+  validate_list(outcome, outcomes)
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply R/cfbd_passing.R:616]: Fixed in `d62438f8` — see the reply on the `R/cfbd_rushing.R` thread for the same issue. Short version: I checked the spec, exactly four of the ten endpoints mark `year` required, and `cfbd_passing_plays()` is one of them, so it now aborts on a NULL year instead of documenting a contract nothing enforced.

File: R/cfbd_rushing.R [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+    },
+    finally = {
+    }
+  )
+  return(df)
+}
+
+#' @title
+#' **Get player game rushing production, split by run direction**
+#' @param year (*Integer* required): Year, 4 digit format (*YYYY*) \cr
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply R/cfbd_rushing.R]: Fixed in `d62438f8`, and the check was worth doing because my docs were wrong in *both* directions.

Against the spec, exactly four of the ten endpoints mark `year` required: `/passing/players/games`, `/passing/teams/games`, `/passing/plays`, `/rushing/teams/games`. Those four now `cli_abort("Missing required field: year")`, so the documented contract is the enforced one.

The other six are documented optional — including `cfbd_rushing_players_games()` and `cfbd_rushing_plays()`, which I had marked *required* purely because they were generated from the passing template, where the corresponding endpoints are. Enforcing that would have rejected valid calls, so there is now a test asserting `/rushing/plays` still accepts a game-scoped call with no year.

File: tests/testthat/test-cfbd_passing_players_season.R [hunk 4]
```diff
@@ -0,0 +1,32 @@
+
+identity_cols <- c("season", "player_id", "player", "team", "conference")
+
+test_that("CFB Passing Players Season", {
+  skip_on_cran()
+  x <- cfbd_passing_players_season(year = 2025, team = "Texas")
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply tests/testthat/test-cfbd_passing_players_season.R:7]: Fixed in `d62438f8`. Every live block now guards on `has_cfbd_key()`; verified by running the suite with the key unset — all 20 blocks skip with "CFBD_API_KEY not set" instead of erroring.

The second half of your point was the sharper one and I had missed it entirely: `expect_error()` with no message assertion passes on `validate_api_key()`s error, so those tests would have gone green while never exercising the enum validation they exist to cover. They now pin the validator message. Worth noting I first wrote those regexes from memory (`"sideways|allowable|one of"`) and they failed — the real text is `Enter valid rush_direction (character): (left, middle, right)`, so the assertions are pinned to that rather than to a guess.

File: vignettes/cfbd_stats.Rmd [hunk 5]
```diff
@@ -131,6 +131,93 @@ passing_df |> gt() |>
...
+  )
+```
+
+Play-level frames expose the same idea as parse flags. Filter on
+`location_analysis_eligible` (passing) or `direction_analysis_eligible`
+(rushing) before trusting a split column:
+
+```{r}
+cfbd_passing_plays(year = 2025, week = 5, outcome = "completion") |>
+  dplyr::filter(!is.na(target)) |>
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply vignettes/cfbd_stats.Rmd]: Fixed in `d62438f8`. The example now filters `location_analysis_eligible, !is.na(target)`, matching the prose immediately above it rather than contradicting it.

File: R/cfbd_passing.R [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+                                       passer_id = NULL,
+                                       classification = NULL) {
+
+  # Validation ----
+  validate_api_key()
+  # The API requires `year` on this endpoint. validate_year(NULL) is a no-op, so
+  # without this the call goes out without the parameter and comes back as an
+  # empty frame with a generic message -- the docs would be promising a contract
+  # nothing enforced (CodeRabbit on #151).
+  if (is.null(year)) cli::cli_abort("Missing required field: year")
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply R/cfbd_passing.R]: Fixed in `52168c33`, and verifying it turned up a wider defect than the report.

Measured first: `year = NA` (logical) and `NA_character_` **already** produced the correct `cli` error. What actually breaks is `NA_integer_`, `NA_real_` and any vector containing `NA`, which give `missing value where TRUE/FALSE needed`.

The root cause is not in this PR. `nchar(NA_integer_)` is `NA`, so `checks` carries an `NA`, `all()` returns `NA`, and `if (!all(checks))` raises a base R error — out of `validate_year()`, the function whose entire job is producing a readable message. That function is shared by **24 files**, so `anyNA(year)` at my four call sites would have fixed the four endpoints you named and left every other caller throwing the same cryptic error. `!all(checks)` -> `!isTRUE(all(checks))` is a smaller diff than four guards and fixes the package.

I did take the `anyNA(year)` guard as well, at the four endpoints where the API marks `year` required — there a `NA` year is genuinely *missing* rather than malformed, so `Missing required field: year` is the more accurate message. Both paths now:

```
required endpoint:  NA / NA_integer_ / NA_character_ / c(2025, NA)
                    -> "Missing required field: year"
optional endpoint:  NA_integer_ / c(2025, NA)
                    -> "Enter valid year as a number (YYYY)"
```

Added `tests/testthat/test-validate_year.R` since the fix now lives in shared code, plus an NA case on `cfbd_passing_plays()`. Ran the **full** package suite rather than just the new files — a shared-validator change tested only by its own caller is how a regression ships — and it is 0 failures across the package.
